sound/soc/sof/ipc4-pcm.c

```diff
@@ -59,19 +59,148 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static void
+sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
+				      struct snd_sof_pipeline *spipe,
+				      struct ipc4_pipeline_set_state_data *trigger_list)
+{
+	struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+	if (pipeline->skip_during_fe_trigger)
+		return;
+
+	switch (state) {
+	case SOF_IPC4_PIPE_RUNNING:
+		/*
+		 * Trigger pipeline if all PCMs containing it are paused or if it is RUNNING
+		 * for the first time
+		 */
+		if (spipe->started_count == spipe->paused_count)
+			trigger_list->pipeline_ids[trigger_list->count++] =
+				pipe_widget->instance_id;
+		break;
+	case SOF_IPC4_PIPE_RESET:
+		/* RESET if the pipeline is neither running nor paused */
+		if (!spipe->started_count && !spipe->paused_count)
```

Comment:
I don't get this, again.
In RUNNING and PAUSED target state you seam to check the started/paused counter _before_ it is adjusted to the new state (so you check the old counters), here it looks to me that you expect that the change has already been done to the counters.
`started_count == 0 && paused_count == 0` is when the pipeline (or pipe?) is already in RESET state, so it is true when there is no stream have started.

There ought to be a simpler way to do this state tracking...
Like use a `current_state` (RESTE/PAUSED/RUNNING) for the 'pipelets' and based on the PCM trigger:
START: started_count++;
1. scan the path for 'pipeliets' which is in RESET state and move them to PAUSED
2. scan again the path and gather the PAUSED ones and move them to RUNNING.

PAUSE: started_count--; paused_count++
1. scan the path for RUNNING && started_count == 0 and move them to PAUSED

PAUSE_RELEASE: stared_count++; paused_count--
1. scan the path for RESET, move them to PAUSED
2. scan the path for PAUSED and move them to RUNNING

STOP: started_count--
1. scan the path for RUNNING && stared_count == 0 and move them to PAUSED
2. scan the path for PAUSED && paused_count == 0, move them to RESET

Reply:
@ujfalusi Im not comfortable with this ref counting for paused/started counts. Do you see a flaw with how it works with my implementation? If not, I'd like to keep it like this. RESET is a special case because it comes after a stop

Reply:
@ujfalusi yes, I have run multiple daily tests with this PR and its an all pass